### PR TITLE
Add custom Unleash Context enrichers

### PR DESCRIPTION
**Pull request by seniorquico**
## About the changes

We use an Unleash SDK in our backend application to evaluate toggles for dynamic feature rollout. In the backend, we sometimes use sensitive properties in our context. This works great since we have full control over the client's context in a trusted environment. However, we weren't sure how to tackle something similar for our JavaScript frontend applications. We'd prefer _very much_ to not send the sensitive properties to the frontend application just so they can be included in the context & calls to the Unleash Proxy. Instead, we thought it would make sense to _enrich_ the context inside the Unleash Proxy. In essence, apply what we do in our backend to the Unleash Proxy.

This PR introduces an extensibility point, similar to loading custom activation strategies, that enables running a series of functions to manipulate the context before calling the Unleash client and resolving the feature toggles. The proposed solution is meant to be fairly generic. We think it could be applied in many ways as an alternative to always needing to assemble the full context in the frontend application.

As an example, consider using one of Maxmind's GeoLite2 databases to lookup geolocations based on the remote IP address and enriching the context with the ISO country code. Here's an example implementation:

```js
const maxmind = require("maxmind");

// This is a Promise that returns the parsed database.
const databaseLoader = maxmind.open("./GeoLite2-Country.mmdb");

function enrichWithCountryCode(context) {
  const remoteAddress = context.remoteAddress;
  if (remoteAddress) {
    return databaseLoader.then((database) => {
      const result = database.get(remoteAddress);
      if (result && result.country && result.country.iso_code) {
        return {
          ...context,
          countryCode: result.country.iso_code,
        };
      }

      return context;
    });
  }

  return Promise.resolve(context);
}

module.exports = [enrichWithCountryCode];
```

To use the above enricher function, one would create a custom Docker image like so:

```Dockerfile
FROM unleashorg/unleash-proxy:latest

ENV UNLEASH_CUSTOM_ENRICHERS_FILE=/unleash-proxy/custom-enrichers.js
ADD --chown=node:node --chmod=0644 ["./custom-enrichers.js", "./GeoLite2-Country.mmdb", "./"]
```

At runtime, the `server.sh` script copies the custom JavaScript file to the `dist` folder. On application startup, the configuration module loads the custom module (if and only if the environment variable/option was set). The Express handler invokes the exported function(s) when a request to resolve feature toggles is received.

### Important files

The `src/custom-enrichers.ts` file contains the new business logic, which is fairly basic. The bulk of the changes exist to plumb the enricher functions into the runtime & Express handler.

## Discussion points

I've opened this PR as a Draft to start a discussion about this potential capability, the team's interest in merging it, and what coding standards should be applied before it's considered for merging. I forked the Unleash Proxy and successfully started to use a modified version with these changes in a production environment. I am happy to help work on this PR and help get it merged if there's the potential others will find it useful.

### CODE REVIEW

**Comment by ivarconr** [comment]
Thanks for sharing 😄 

This is something we have been thinking about previously and your proposal is interesting. I will discuss this internally with the team and come back to you! 🚀

**Comment by ivarconr** [comment]
Sorry for the slow turn-around on my end. We are willing to explore this feature, but will wrap it in experimental flags for now, to allow us to change the design somewhat in the future. 

Our biggest concern at the moment is that there can be strange things going on inside the enriches (calling slow/unstable third party services) that can significantly affect the performance and stability of the proxy itself. 

We have also discussed a slightly different approach, where the resolves works as "middlewares" and enrich the request on the way in, instead of adding it in the resolve path itself.

**Comment by seniorquico** [comment]
No worries on the time needed to review this! As I noted, we've been running with a forked build based on this branch and it's working very well.

I understand the concern around slowing down request processing. But isn't this a concern for the implementer? I'd assume someone extending the proxy would need to make careful considerations about what they put inline. For our use case, we make extensive use of caching to ensure the decision introduces as little delay as possible. We've not observed any significant difference in average response times and we've observed negligible increase in memory utilization of the proxy server. But I'd think that all to be highly dependent on the implementation.

How best do you propose proceeding? I'm not at all stuck on this specific implementation approach (chained requests in the resolve method). It simply seemed like a relatively straightforward/easy way of getting something up and running. I'd be happy to help pitch in with some follow-on development work here. We're finding great value in the overall concept. We've since rolled in a few additional context properties for our users, and our product team is _loving_ the new flexibility it's providing in targeting users.

**Comment by ivarconr** [comment]
> How best do you propose proceeding?

For now, only thing would be rename the option to be an experimental option ([see examples](https://github.com/Unleash/unleash-proxy#experimental-options)) and we can merge this one.

**Comment by seniorquico** [comment]
I renamed the option to `EXP_CUSTOM_ENRICHERS_FILE`. Let me know if there are any other items to consider.

File: server.sh [hunk 1]
```diff
@@ -9,4 +9,11 @@ if [ ! -z "$UNLEASH_CUSTOM_STRATEGIES_FILE" ]; then
 	export UNLEASH_CUSTOM_STRATEGIES_FILE="./strategies/$(basename $UNLEASH_CUSTOM_STRATEGIES_FILE)"
 fi
 
+if [ ! -z "$EXP_CUSTOM_ENRICHERS_FILE" ]; then
+	echo "Loading custom enrichers from file $EXP_CUSTOM_ENRICHERS_FILE"
+	mkdir -p ./dist/enrichers
+	cp $EXP_CUSTOM_ENRICHERS_FILE ./dist/enrichers
+	export EXP_CUSTOM_ENRICHERS_FILE="./enrichers/$(basename $EXP_CUSTOM_ENRICHERS_FILE)"
```
[the remark this answers is not in the document]
Comment (gastonfournier) [reply server.sh:16]:
```suggestion
	export EXP_CUSTOM_ENRICHERS_FILE="./enrichers/$(basename "$EXP_CUSTOM_ENRICHERS_FILE")"
```
Something like this should work based on this testing:
```bash
$ export EXP_CUSTOM_ENRICHERS_FILE=/tmp/test\ with\ spaces/vote.png
$ echo "./enrichers/$(basename $EXP_CUSTOM_ENRICHERS_FILE)"
basename: extra operand ‘spaces/vote.png’
Try 'basename --help' for more information.
./enrichers/
$ echo "./enrichers/$(basename "$EXP_CUSTOM_ENRICHERS_FILE")"
./enrichers/vote.png
```

File: server.sh [hunk 2]
```diff
@@ -9,4 +9,11 @@ if [ ! -z "$UNLEASH_CUSTOM_STRATEGIES_FILE" ]; then
 	export UNLEASH_CUSTOM_STRATEGIES_FILE="./strategies/$(basename $UNLEASH_CUSTOM_STRATEGIES_FILE)"
 fi
 
+if [ ! -z "$EXP_CUSTOM_ENRICHERS_FILE" ]; then
+	echo "Loading custom enrichers from file $EXP_CUSTOM_ENRICHERS_FILE"
+	mkdir -p ./dist/enrichers
+	cp $EXP_CUSTOM_ENRICHERS_FILE ./dist/enrichers
```
[the remark this answers is not in the document]
Comment (seniorquico) [reply server.sh]: Fixed!

File: server.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (seniorquico) [reply server.sh]: Fixed!

File: server.sh [hunk 3]
```diff
@@ -9,4 +9,11 @@ if [ ! -z "$UNLEASH_CUSTOM_STRATEGIES_FILE" ]; then
 	export UNLEASH_CUSTOM_STRATEGIES_FILE="./strategies/$(basename $UNLEASH_CUSTOM_STRATEGIES_FILE)"
 fi
 
+if [ ! -z "$EXP_CUSTOM_ENRICHERS_FILE" ]; then
```
[the remark this answers is not in the document]
Comment (seniorquico) [reply server.sh]: Fixed!

File: README.md [hunk 4]
```diff
@@ -141,12 +141,14 @@ Keep-Alive: timeout=5
...
+| expBootstrap.urlHeaders.Authorization | `EXP_BOOTSTRAP_AUTHORIZATION`       | n/a           |    no    | Authorization header value to be used when bootstrapping                                                                                                                                                                                                                       |
+| expServerSideSdkConfig.tokens         | `EXP_SERVER_SIDE_SDK_CONFIG_TOKENS` | n/a           |    no    | API tokens that can be used by Server SDKs (and proxies) to read feature toggle configuration from this Proxy instance.                                                                                                                                                        |
+| expCustomEnrichers                    | `EXP_CUSTOM_ENRICHERS_FILE`         | []            |    no    | Use this option to inject implementation of custom context enrichers. If you are using `UNLEASH_EXP_CUSTOM_ENRICHERS_FILE` you need to provide a valid path to a javascript files which exports an array of custom context enrichers and the SDK will automatically load these |
```
[the remark this answers is not in the document]
Comment (seniorquico) [reply README.md]: Fixed!

**Comment by seniorquico** [comment]
@gastonfournier I fixed the SonaType warnings and resolved the merge conflict. Let me know if you see anything additional!

**Review by sighphyre — APPROVED** [review]
I think this looks good now, it's going in!

**Comment by LukasHeimann** [comment]
Hi everyone,

thank you for this really cool new feature! In my project, we have been manipulating the `req` before the proxy got to process it, which was always a bit awkward as we have to know when and how it is parsed to the context. However, we rely on information at the `req` to process "enrich" our context. Would it be possible to pass the request to the `ContextEnricher` as well?

Thanks
Lukas

**Pull request merged**